drivers/sensor/lsm6dso16is/lsm6dso16is_shub.c

```diff
@@ -0,0 +1,838 @@
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/*
+ * use SLV0 for generic read to device
+ */
+static int lsm6dso16is_shub_read_dev_reg(const struct device *dev,
+					 uint8_t slv_addr, uint8_t slv_reg,
+					 uint8_t *value, uint16_t len)
+{
+	const struct lsm6dso16is_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	uint8_t device[3];
+
+	device[0] = (slv_addr << 1) | LSM6DSO16IS_SHUB_SLVX_READ;
+	device[1] = slv_reg;
+	device[2] = (len & 0x7);
+
+	if (lsm6dso16is_shub_write_embedded_regs(ctx, LSM6DSO16IS_SHUB_SLV0_ADDR,
+					     device, 3) < 0) {
+		LOG_DBG("shub: error writing embedded reg");
+		return -EIO;
+	}
+
+	/* turn SH on, wait for shub i2c read to finish */
+	lsm6dso16is_shub_enable(dev, 1);
+	lsm6dso16is_shub_wait_completed(ctx);
+
+	/* read data from external dev */
+	lsm6dso16is_shub_embedded_en(ctx, true);
+	if (lsm6dso16is_read_reg(ctx, LSM6DSO16IS_SHUB_DATA_OUT,
+			     value, len) < 0) {
+		LOG_DBG("shub: error reading sensor data");
+		lsm6dso16is_shub_embedded_en(ctx, false);
+		return -EIO;
+	}
+	lsm6dso16is_shub_embedded_en(ctx, false);
+
+	if (lsm6dso16is_shub_check_slv0_nack(ctx) < 0) {
+		lsm6dso16is_shub_enable(dev, 0);
+		return -EIO;
+	}
+
+	lsm6dso16is_shub_enable(dev, 0);
+	return 0;
+}
+
+/*
+ * use SLV0 to configure  device
+ */
+static int lsm6dso16is_shub_write_dev_reg(const struct device *dev,
+					  uint8_t slv_addr, uint8_t slv_reg,
+					  uint8_t *value, uint16_t len)
+{
+	const struct lsm6dso16is_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	uint8_t slv_cfg[3];
+	uint8_t cnt = 0U;
+
+	while (cnt < len) {
+		slv_cfg[0] = (slv_addr << 1) & ~LSM6DSO16IS_SHUB_SLVX_READ;
+		slv_cfg[1] = slv_reg + cnt;
+
+		if (lsm6dso16is_shub_write_embedded_regs(ctx,
+						     LSM6DSO16IS_SHUB_SLV0_ADDR,
+						     slv_cfg, 2) < 0) {
+			LOG_DBG("shub: error writing embedded reg");
+			return -EIO;
+		}
+
+		slv_cfg[0] = value[cnt];
+		if (lsm6dso16is_shub_write_embedded_regs(ctx,
+					LSM6DSO16IS_SHUB_SLV0_DATAWRITE,
+					slv_cfg, 1) < 0) {
+			LOG_DBG("shub: error writing embedded reg");
+			return -EIO;
+		}
+
+		/* turn SH on, wait for shub i2c write to finish */
+		lsm6dso16is_shub_enable(dev, 1);
+		lsm6dso16is_shub_wait_completed(ctx);
+
+		if (lsm6dso16is_shub_check_slv0_nack(ctx) < 0) {
+			lsm6dso16is_shub_enable(dev, 0);
+			return -EIO;
+		}
+
+		lsm6dso16is_shub_enable(dev, 0);
+
+		cnt++;
+	}
+
+	/* Put SLV0 in IDLE mode */
+	slv_cfg[0] = 0x7;
+	slv_cfg[1] = 0x0;
+	slv_cfg[2] = 0x0;
+	if (lsm6dso16is_shub_write_embedded_regs(ctx, LSM6DSO16IS_SHUB_SLV0_ADDR,
+					     slv_cfg, 3) < 0) {
+		LOG_DBG("shub: error writing embedded reg");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/*
+ * SLAVEs configurations:
+ *
+ *  - SLAVE 0: used for configuring all devices
+ *  - SLAVE 1: used as data read channel for external device #1
+ *  - SLAVE 2: used as data read channel for external device #2
+ *  - SLAVE 3: used for generic reads while data channel is enabled
```

Comment:
slave

Reply:
yes, master! (hope I can joke a bit about it... :D ) OK, I'll take care of it. Sorry, I just looked for the word "slave" in lowe case.